File: ws/robalino/u1/WS_05_DataPersistance/javascriptcode/dataManager.js
// File: dataManager.js
const fs = require('fs');
const Person = require('./person');

class DataManager {
    constructor(filePath) {
        this.filePath = filePath;
        this.people = this.loadFromFile();
    }

    loadFromFile() {
        try {
            const data = fs.readFileSync(this.filePath, 'utf8');
            const peopleArray = JSON.parse(data);
            return peopleArray.map(person => Person.fromJSON(person));
        } catch (error) {
            if (error.code === 'ENOENT') {
                return []; // File does not exist, return an empty list
            }
            throw error;
        }
    }

    saveToFile() {
        const data = JSON.stringify(this.people.map(person => person.toJSON()), null, 4);
        fs.writeFileSync(this.filePath, data, 'utf8');
    }

    addPerson(person) {
        this.people.push(person);
        this.saveToFile();
    }

    updatePerson(id, updatedPerson) {
        this.people = this.people.filter(person => person.id !== id); // Remove the old person
        this.addPerson(updatedPerson); // Add the updated person
    }

    deletePerson(id) {
        this.people = this.people.filter(person => person.id !== id);
        this.saveToFile();
    }

    searchById(id) {
        return this.people.find(person => person.id === id) || null;
    }

    getAllPeople() {
        return this.people;
    }
}

module.exports = DataManager;

File: ws/robalino/u1/ws05dataPersistance/pythoncode/data_manager.py
# File: data_manager.py
import json
from person import Person

class DataManager:
    def __init__(self, file_path):
        self.file_path = file_path
        self.people = self.load_from_file()

    def load_from_file(self):
        try:
            with open(self.file_path, 'r') as file:
                data = json.load(file)
                return [Person.from_dict(person) for person in data]
        except FileNotFoundError:
            return []  # Start with an empty list if the file does not exist

    def save_to_file(self):
        with open(self.file_path, 'w') as file:
            json.dump([person.to_dict() for person in self.people], file)

    def add_person(self, person):
        self.people.append(person)
        self.save_to_file()

    def update_person(self, person_id, updated_person):
        self.people = [p for p in self.people if p.id != person_id]  # Remove existing person
        self.add_person(updated_person)

    def delete_person(self, person_id):
        self.people = [p for p in self.people if p.id != person_id]
        self.save_to_file()

    def search_by_id(self, person_id):
        for person in self.people:
            if person.id == person_id:
                return person
        return None

    def get_all_people(self):
        return self.people
